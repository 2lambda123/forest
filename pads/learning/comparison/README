This directory stores the results from golden and silver IRs for a number of
selected examples. Currently the examples being tested are:

1967Transactions.short
ai.3000
boot.txt
crashreporter.log
dibbler.1000
ls-l.txt
quarterlypersonalincome.txt
railroad.txt
rpmpkgs.txt
yum.txt

The results from the golden IRs are saved in files with an extension ".golden";
the results from the solver IRs are saved in files with an extension ".solver".

This file will be a placeholder for future comments about the comparisons.

==============================================================================
dibbler.1000 thoughts
==============================================================================
(not all these thoughts pertain directly to dibbler.1000)

1. there is a rough structure processing step which should trigger if
it sees a histogram with a variable number of occurrences of a
particular token, but that variable number of occurrences is always
greater than some fixed number, say N.  What it should do is chop the
the input into two halves -- the first half with the fixed initial N
tokens and the second half with a variable number of tokens.  This
does not seem to be happening in the dibbler.1000 example.  Perhaps
because of the single header record that screws it up? Perhaps the
decision needs to be done on an approximate basis -- do it if there
are only one or two records that deviate from the pattern; throw
the deviant records into another case in a sum?

2. I know Kenny has worked hard at Pempty elimination, but I think the
following transformation, which uses a Pempty (print as PADS/C Pvoid)
is a good one:

Poption [ Union [ field1; field2; ...; fieldk ] ]

==>

Union [ field1; field2; ...; fieldk; Pempty fieldname]

It is like flattening nested unions into a single union.

I think that this should have the tiniest decrease in complexity
because in the first description, a Poption will have a (log 2) factor
+ a (log k) factor (the latter from the number of branches in the
union; the former from the fact that Poption is a union with two
branches).  Whereas in the second description, there will only be a
(log (k+1)) factor which is a tiny bit smaller than (log 2) + (log k).

3. it would be nice to have a compiler flag that did not print out the
complexity information.

4. my perception of how well the inference algorithm does may be
influenced by formatting.  a more-compact-looking description gives me
more satisfaction than a less-compact-looking description, even when the
descriptions are semantically the same.  in the future, it would be nice to
be able to print out descriptions in PADS/ML syntax.  In particular,
the presence of anonymous tuples would decrease the "apparent" complexity
of types (though making no impact on the "real" complexity) -- since label
names are meaningless anyway (except where there are dependencies), eliminating
simplifies the presentation.

Note the beauty of:

-----
pdatatype elem =
  Field0 of (Pstring * White) option * Pint
| Field1 of  Pstring * (White * Pstring) option

ptype record = elem option Plist('\n','|')
-----

vs this (which isn't PADS/C syntax, but is closer to the PADS/C level of
 verboseness):

-----
RArray
	Separator: [StringConst] "|"
	Poption
		Punion
			Pstruct
				Poption
					Pstruct
						[String] 
						[White] 
					End Pstruct;
				End Poption;
				[Pint] 
			End Pstruct;
			Pstruct
				[String] 
				Poption
					Pstruct
						[White]
						[String]
					End Pstruct;
				End Poption;
			End Pstruct;
		End Punion;
	End Poption;
End RArray
-----

==========================================================================
ls -l thoughts
=============================================================================

1. briefly: 
I started looking at the ls-l data.  it seems that there is not enough data.  It looks over-fit to me.  I wonder if we can define a metric that avoids overfitting by not adding dependencies/enums/switches when there is not a "statistically significant" enum/dependency hypothesis.

