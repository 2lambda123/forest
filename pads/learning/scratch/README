This directory contains a prototype system for inferring a desription
of newline-separated ASCII data.  

Requirements:
You must have sml/nj 110.59 or later.

To build:
> make

To invoke:
./learn rawfile outputdir
   rawfile   contains the data to run the inference on
   outputdir names the directory for ouptut

The tool currently sends to standard out the following information:
* A count of the number of records in the input file
* A list of the frequency distributions of the tokens in the file,
  grouped into clusters, sorted by the cluster with the lowest
  "score".  The score is a heuristic designed to find the cluster with
  the strongest signal.  
* An output indicating which cluster the system chose and the coverage
  of that cluster. 
* A representation of the inferred top-level type of the data.
  Contexts that have not yet been analyzed are marked TBD.  For each
  such marker TBD_i in the type representation, there is a file in
  outputdir/TBD_i containing the portion of the original input that
  falls into that context.
