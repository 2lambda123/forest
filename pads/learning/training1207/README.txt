How to define a training description:

    1)  Only use base types listed below & literal punctuation.
             -- eg: no Pstring
    2)  Define the description that you would like the tokenizer to infer 
          (not necessarily the description that you hope the system will
           eventually output)

question:  what happens to unseparated literals?  eg: " ["
question: do we need Pignore when you can't figure out what to do?
question: what happens to array separators (do they get printed?)
q: should time & date tokens be separated?

Token Definitions 
from Qian's Email re 11/16/2007 12:56 AM:

Meta-Rule:
-- Tokenization shouldn't break nesting of well-nested parens or quotes. 
-- ie: ( ) [ ] { } " "  ' ' < >

Pint		|	numbers									
Pfloat		|	numbers "." numbers 
Ptimez		|	2 digits ":" 2 digits ":" 2 digits, followed by an option of "am" or "pm", and an option of timezone
Pip		|	1-3 digits "." 1-3 digits "." 1-3 digits "." 1-3 digits	
Phostname	|	a series of labels, with each label being separated by "." (restrictions on valid host names can be found at http://en.wikipedia.org/wiki/Hostname)
Pemail		|	any strings that may have numbers, letters, "_" and "-", but doesn't begin with "_" or "-", followed by "@", followed by hostname
Pmac		|	6 2-digit hex numbers separated by ":" or "-"
Pdate		|	any strings containing numbers and letters that represent a date
Ppath		|	a series of filename separated by "/" or "\" (filename limitations can be found at http://en.wikipedia.org/wiki/Filename)
Purl		|	protocal string, hostname, filename or ip separated by various punctuations
Pword		|	strict English world containing letters and "'"
Pid		|	any strings that are composed of numbers, letters, "-", "_" and "."
PbXML		|	XML data token
PeXML		|	another XML data token
Pwhite		|	white spaces
Pmessage        |       any characters to end of line or delimiting scope like ""?  Takes
                        terminating character as argument.

Pdot		|	"."
Pslash		|	"/"
Psemicolon	|	";"
Pbar		|	"|"
Pless		|	"<"
.
.
.
(This is a group of punctuation tokens, with each punctuation being a unique token)

Pblob		|	any strings that can't be represented by the above tokens, but must have a terminated character, like "\n" or "\r" 