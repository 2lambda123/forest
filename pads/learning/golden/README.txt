Suggested Order:

rpmpkgs.txt	3 hrs
quarterlypersonalincome.txt 2 days
yum.txt 5 hrs
boot.txt 1 hr
convert.log 12 hrs
ls-l.txt 1 hr
1967Transactions.short 4 hrs
ai.3000 1 hr
ai.big    -- same format as ai.3000 but longer
dibbler.1000 0.5 hr
dibbler.10001  -- same format as dibbler.1000 but longer
